### ISSUE TITLE
[Question] Is it possible to encrypt DRM protected content with a 'clear lead'

### ISSUE THREAD

**Issue by kidmiracleman** [axiomatic-systems/Bento4#63]
Hi,

I am trying to use the Bento packager (mp4-dash.py) to encrypt content with a clear lead.

By 'clear lead' I am referring to the ability to leave certain  NAL types in the clear (if present at beginning ) e.g RAP/IDR's  

Is this possible with he current tool version?

**Comment by barbibulle** [comment]
That's already the case by default. When encrypting through mp4-dash.py, what's implemented is MPEG Common Encryption as specified in ISO/IEC DIS 23001-7 3rd Edition, which means that for H.264 and H.265, only VCL NAL Units are encrypted, and those that are encrypted keep the NAL header as well as slice headers in the clear.

**Comment by kidmiracleman** [comment]
Hi Gilles,

Is there an easy way to verify this at the container level?

I'm looking at issues with DASH/Widevine playback on specific Android devices and I can only reproduce with Bento packaged content   -  when using MP4Box explicitly set not to encrypt RAPs/IDRs  I never get issues.  This is why I am looking at trying to ascertain exact differences in how encryption is applied (since I can't change the client side)

**Comment by barbibulle** [comment]
There's a known issue with some intel-based Android devices where the decrypter supplied by Intel doesn't like partially-encrypted NAL units, even though that's what the spec requires. More specifically, that decrypter will only accept samples either fully encrypted, fully in the clear, or with just the first 5 bytes of the NAL unit in the clear (size and type). It would be possible to support this mode in the Bento4 encrypter, but that would create a non-compliant bitstream, so it is problematic.
Can you share one of the streams you generated with MP4Box so that I can check whether this is indeed the difference here?
If that's the case, and it really isn't possible to obtain a fix for the client decrypter, I could always add an option to the code to allow for that mode, with the caveat that not all other decrypters are guaranteed to accept it.

**Comment by kidmiracleman** [comment]
Hi,

You have guessed right RE: Android/Intel decoder.  Its actually only the secure decoder that is effected the s/w video path seems to work OK.

My earlier comment seems to be wrong  - I *Think* that the MP4Box content is fully encrypted by default - here is a link :  https://dl.dropboxusercontent.com/u/25125951/gpac_full_enc.tar

Also - the YouTube Widevine test vectors seem to work : http://www.youtube.com/api/manifest/dash/id/d286538032258a1c/source/youtube?as=fmp4_audio_cenc,fmp4_sd_hd_cenc&sparams=ip,ipbits,expire,source,id,as&ip=0.0.0.0&ipbits=0&expire=19000000000&signature=477CF7D478BE26C205045D507E9358F85F84C065.8971631EB657BC33EC2F48A2FF4211956760C3E9&key=ik0

Currently we use Bento as a reference tool to demo multiDRM CENC content security  - so I'm really only trying to find out what the critical factor here is.  I don't think modifying the tool to produce a non-compliant bitstream is something we would need currently.

Thanks for you help thus far!

**Comment by barbibulle** [comment]
This MP4Box-encrypted content has the first 5 bytes of each NAL unit (NAL unit size and NAL unit type) in the clear, followed by encrypted bytes. This isn't compliant with the latest version of the spec.

**Comment by kidmiracleman** [comment]
Thank You Gilles,

I understand now (I think)

Older spec  - just leave NAL unit size and type  in clear and encrypt the rest (if VCL NAL)?
Latest spec  - VCL NAL unit will have 'interleaved' encryption   -with some portions of the bitstream below the NAL header left in the clear?

Is the a tool I can use to verify this at the codec level ?  or can I just parse the SENC box to infer this?

Please close this ticket if you want.

Thanks a lot!

**Comment by barbibulle** [comment]
You can use mp4dump --verbosity 3 <filename> to show the contents of the sample encryption table (located in the 'senc' box).

**Comment by kidmiracleman** [comment]
Thanks,

Thats a great help.

Please close this issue.

**Comment by kidmiracleman** [comment]
Hi,

Is it possible to re-open this issue so that you can consider (configurable) support for eh older version of the specification?

From further testing it appears that the latest spec demonstrates many compatibility issues with legacy Android devices.  This is not surprising since the OEMs do not maintain updates for these devices for very long...   The Nexus 6P (new device) also has issues with this version of the spec.

**Comment by barbibulle** [comment]
Good news: I have implemented an option that allows the encrypter to follow the "old" layout. I'll push a version of the code with this option sometime this week.

**Comment by kidmiracleman** [comment]
Thank you.

**Comment by mlevkov** [comment]
Is there is a way to have a "clear lead" of configurable time or set time, for example, 2 seconds, for HLS packaged content using Bento4?

**Comment by mlevkov** [comment]
Hello, 
I'm open to helping you with testing of this feature ;).
Would you please help me understand how this attribute is set within the command line interface. 
Am I able to use this with hls4-hls_py? Are attributes set by [type=audio, ClearLeadFragments=2]video.mp4 or in some other form?

Could you please provide a sample command line that you expect mp4encrypt for Fairplay SAMPLE encryption?

From what I understand, the mp4encrypt is geared towards DASH packaging and not HLS, am I correct

**Comment by barbibulle** [comment]
You're right, the mechanism I describe here only works for MP4 files (and as a result with DASH).
For HLS, this isn't currently implemented. It is fairly simple to add, so I will schedule this feature for the next release.
With MP4 files, you would set the property like this, for 2 cleartext leading segments (in this example, the file has two tracks, ID 1 and 2, and we're using the same KID and key for both tracks):
`mp4encrypt --method MPEG-CENC --property 1:ClearLeadFragments:2 --property 2:ClearLeadFragments:2 --key 1:000102030405060708090a0b0c0d0e0f:random --key 2:000102030405060708090a0b0c0d0e0f:random --property 1:KID:00112233445566778899aabbccddeeff --property 2:KID:00112233445566778899aabbccddeeff video.mp4 encrypted_with_clearlead.mp4`

If you're using mp4dash, the KID and encryption keys are passed for you to mp4encrypt internally, so all you need to specify is the extra parameters (you'll still need to be explicit about the track IDs):
`mp4dash --encryption-args="--property 1:ClearLeadFragments:2 --property 2:ClearLeadFragments:2" --encryption-key=00112233445566778899aabbccddeeff:000102030405060708090a0b0c0d0e0f video.mp4`

**Comment by mlevkov** [comment]
Thank you for the sample and detailed command line structure. I will run some tests with our data and let you know what comes out of it. For the HLS/Fairplay effort, what are your thoughts on the timing for next release?

Also, do you have a sample structure for the final outcome. From what I understand, it is a subvariant implementation, not a master playlist change. In the subvariant, the X-EXT-KEY:METHOD=SAMPLE-AES would be shifted by the segments in the clear (unencrypted),
as such:
....
#EXT-X-TARGETDURATION:4
#EXT-X-MEDIA-SEQUENCE:0
#EXTINF:4.004004,
segment_0.ts
#EXT-X-KEY:METHOD=SAMPLE-AES,<other key details go here...>
#EXTINF:4.004004,
segment_1.ts

where segment_0.ts is not encrypted and any segments following the #EXT-X-KEY:METHOD=SAMPLE-AES would have to be encrypted.
Or #EXT-X-KEY:METHOD=NONE would have to be placed before #EXTINF:4.004004,
segment_0.ts of segment in the clear.

I'm confident that you know how the proper implementation needs to look like, I'm just curious of the specifics a bit in advance ;). Also, would you be kind to share some thoughts on what your timing is for the next release?
